File: packages/ios/ContentfulOptimization/Sources/ContentfulOptimization/Core/OptimizationConfig.swift
import Foundation

private func normalizeLocale(_ locale: String?) -> String? {
    guard let locale else { return nil }

    let normalizedLocale = locale.trimmingCharacters(in: .whitespacesAndNewlines)
        .replacingOccurrences(of: "_", with: "-")
    let matchKey = normalizedLocale.lowercased()

    guard !normalizedLocale.isEmpty,
          normalizedLocale != "*",
          matchKey != "und",
          normalizedLocale.range(
            of: #"^[A-Za-z]{2,8}(?:-[A-Za-z0-9]{1,8})*$"#,
            options: .regularExpression
          ) != nil
    else { return nil }

    return normalizedLocale
}

private func normalizeExplicitLocale(_ locale: String?, name: String) throws -> String {
    guard let normalized = normalizeLocale(locale) else {
        throw OptimizationError.configError("\(name) must be a valid locale string")
    }

    return normalized
}

/// Defaults that can be restored from persistent storage.
public struct StorageDefaults {
    public var consent: Bool?
    public var persistenceConsent: Bool?
    public var profile: [String: Any]?
    public var changes: [[String: Any]]?
    public var personalizations: [[String: Any]]?

    public init(
        consent: Bool? = nil,
        persistenceConsent: Bool? = nil,
        profile: [String: Any]? = nil,
        changes: [[String: Any]]? = nil,
        personalizations: [[String: Any]]? = nil
    ) {
        self.consent = consent
        self.persistenceConsent = persistenceConsent
        self.profile = profile
        self.changes = changes
        self.personalizations = personalizations
    }
}

/// Configuration for initializing the Contentful Optimization SDK.
public struct OptimizationConfig {
    public let clientId: String
    public let environment: String
    public let experienceBaseUrl: String?
    public let insightsBaseUrl: String?
    /// Default SDK locale used for Experience API requests and event context.
    public let locale: String?
    public var defaults: StorageDefaults?
    public let allowedEventTypes: [String]?

    /// When `true`, the SDK emits detailed diagnostic logs via `os.Logger`
    /// under the subsystem `com.contentful.optimization`.
    /// Logs are visible in Xcode console and Console.app.
    public var debug: Bool

    public init(
        clientId: String,
        environment: String = "master",
        experienceBaseUrl: String? = nil,
        insightsBaseUrl: String? = nil,
        locale: String? = nil,
        defaults: StorageDefaults? = nil,
        debug: Bool = false,
        allowedEventTypes: [String]? = nil
    ) {
        self.clientId = clientId
        self.environment = environment
        self.experienceBaseUrl = experienceBaseUrl
        self.insightsBaseUrl = insightsBaseUrl
        self.locale = locale
        self.defaults = defaults
        self.debug = debug
        self.allowedEventTypes = allowedEventTypes
    }

    /// Normalizes the SDK locale for Experience API requests and event context.
    public func normalizedLocale() throws -> String? {
        guard let locale else { return nil }

        return try normalizeExplicitLocale(locale, name: "locale")
    }

    /// Serializes config to a JSON string for passing to the JS bridge.
    func toJSON(
        anonymousId: String? = nil
    ) throws -> String {
        var dict: [String: Any] = [
            "clientId": clientId,
            "environment": environment,
        ]
        if let url = experienceBaseUrl {
            dict["experienceBaseUrl"] = url
        }
        if let url = insightsBaseUrl {
            dict["insightsBaseUrl"] = url
        }
        if let locale = try normalizedLocale() {
            dict["locale"] = locale
        }
        if let allowedEventTypes {
            dict["allowedEventTypes"] = allowedEventTypes
        }
        if defaults != nil || anonymousId != nil {
            var defaultsDict: [String: Any] = [:]
            if let consent = defaults?.consent {
                defaultsDict["consent"] = consent
            }
            if let persistenceConsent = defaults?.persistenceConsent {
                defaultsDict["persistenceConsent"] = persistenceConsent
            }
            if let profile = defaults?.profile {
                defaultsDict["profile"] = profile
            }
            if let changes = defaults?.changes {
                defaultsDict["changes"] = changes
            }
            if let personalizations = defaults?.personalizations {
                defaultsDict["optimizations"] = personalizations
            }
            if let anonymousId {
                defaultsDict["anonymousId"] = anonymousId
            }
            if !defaultsDict.isEmpty {
                dict["defaults"] = defaultsDict
            }
        }

        let data = try JSONSerialization.data(withJSONObject: dict)
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}

File: packages/android/ContentfulOptimization/src/main/kotlin/com/contentful/optimization/core/OptimizationConfig.kt
package com.contentful.optimization.core

import org.json.JSONObject

private fun normalizeLocale(locale: String?): String? =
    locale
        ?.trim()
        ?.replace('_', '-')
        ?.takeUnless { it.isBlank() || it == "*" || it.lowercase() == "und" }
        ?.takeIf { LOCALE_REGEX.matches(it) }

private val LOCALE_REGEX = Regex("^[A-Za-z]{2,8}(?:-[A-Za-z0-9]{1,8})*$")

private fun normalizeExplicitLocale(locale: String?, name: String): String =
    normalizeLocale(locale) ?: throw OptimizationError.ConfigError("$name must be a valid locale string")

public data class StorageDefaults(
    var consent: Boolean? = null,
    var persistenceConsent: Boolean? = null,
    var profile: Map<String, Any>? = null,
    var changes: List<Map<String, Any>>? = null,
    var selectedOptimizations: List<Map<String, Any>>? = null,
)

public data class OptimizationApiConfig(
    val experienceBaseUrl: String? = null,
    val insightsBaseUrl: String? = null,
    val enabledFeatures: List<String>? = null,
    val preflight: Boolean? = null,
) {
    internal fun isEmpty(): Boolean =
        experienceBaseUrl == null &&
            insightsBaseUrl == null &&
            enabledFeatures == null &&
            preflight == null

    internal fun toJSONObject(): JSONObject {
        val obj = JSONObject()
        experienceBaseUrl?.let { obj.put("experienceBaseUrl", it) }
        insightsBaseUrl?.let { obj.put("insightsBaseUrl", it) }
        enabledFeatures?.let { obj.put("enabledFeatures", org.json.JSONArray(it)) }
        preflight?.let { obj.put("preflight", it) }
        return obj
    }
}

public enum class OptimizationLogLevel(public val wireValue: String) {
    fatal("fatal"),
    error("error"),
    warn("warn"),
    info("info"),
    debug("debug"),
    log("log"),
}

public data class QueueFlushPolicy(
    val flushIntervalMs: Int? = null,
    val baseBackoffMs: Int? = null,
    val maxBackoffMs: Int? = null,
    val jitterRatio: Double? = null,
    val maxConsecutiveFailures: Int? = null,
    val circuitOpenMs: Int? = null,
) {
    internal fun isEmpty(): Boolean =
        flushIntervalMs == null &&
            baseBackoffMs == null &&
            maxBackoffMs == null &&
            jitterRatio == null &&
            maxConsecutiveFailures == null &&
            circuitOpenMs == null

    internal fun toJSONObject(): JSONObject {
        val obj = JSONObject()
        flushIntervalMs?.let { obj.put("flushIntervalMs", it) }
        baseBackoffMs?.let { obj.put("baseBackoffMs", it) }
        maxBackoffMs?.let { obj.put("maxBackoffMs", it) }
        jitterRatio?.let { obj.put("jitterRatio", it) }
        maxConsecutiveFailures?.let { obj.put("maxConsecutiveFailures", it) }
        circuitOpenMs?.let { obj.put("circuitOpenMs", it) }
        return obj
    }
}

public data class QueuePolicy(
    val flush: QueueFlushPolicy? = null,
    val offlineMaxEvents: Int? = null,
    val onOfflineDrop: ((QueueEvent) -> Unit)? = null,
    val onFlushFailure: ((QueueEvent) -> Unit)? = null,
    val onCircuitOpen: ((QueueEvent) -> Unit)? = null,
    val onFlushRecovered: ((QueueEvent) -> Unit)? = null,
) {
    internal fun isEmpty(): Boolean =
        (flush == null || flush.isEmpty()) && offlineMaxEvents == null

    internal fun toJSONObject(): JSONObject {
        val obj = JSONObject()
        if (flush != null && !flush.isEmpty()) {
            obj.put("flush", flush.toJSONObject())
        }
        offlineMaxEvents?.let { obj.put("offlineMaxEvents", it) }
        return obj
    }
}

public data class BlockedEvent(
    val reason: String,
    val method: String,
    val args: List<Any>,
)

public enum class QueueEventType(public val wireValue: String) {
    offlineDrop("offlineDrop"),
    flushFailure("flushFailure"),
    circuitOpen("circuitOpen"),
    flushRecovered("flushRecovered");

    public companion object {
        public fun fromWireValue(value: String): QueueEventType? =
            values().firstOrNull { it.wireValue == value }
    }
}

public data class QueueEvent(
    val type: QueueEventType,
    val context: Map<String, Any>,
)

public data class OptimizationConfig(
    val clientId: String,
    val environment: String = "main",
    val api: OptimizationApiConfig? = null,
    /** Default SDK locale used for Experience API requests and event context. */
    val locale: String? = null,
    var defaults: StorageDefaults? = null,
    val allowedEventTypes: List<String>? = null,
    val logLevel: OptimizationLogLevel = OptimizationLogLevel.error,
    val queuePolicy: QueuePolicy? = null,
    val onEventBlocked: ((BlockedEvent) -> Unit)? = null,
) {
    fun normalizedLocale(): String? =
        locale?.let { normalizeExplicitLocale(it, "locale") }

    internal fun toJSON(anonymousId: String? = null): String {
        val obj = JSONObject()
        obj.put("clientId", clientId)
        obj.put("environment", environment)
        obj.put("logLevel", logLevel.wireValue)
        if (api != null && !api.isEmpty()) {
            obj.put("api", api.toJSONObject())
        }
        normalizedLocale()?.let { obj.put("locale", it) }
        allowedEventTypes?.let { obj.put("allowedEventTypes", org.json.JSONArray(it)) }
        if (queuePolicy != null && !queuePolicy.isEmpty()) {
            obj.put("queuePolicy", queuePolicy.toJSONObject())
        }

        if (defaults != null || anonymousId != null) {
            val defaultsObj = JSONObject()
            val d = defaults
            d?.consent?.let { defaultsObj.put("consent", it) }
            d?.persistenceConsent?.let { defaultsObj.put("persistenceConsent", it) }
            d?.profile?.let { defaultsObj.put("profile", JSONObject(it)) }
            d?.changes?.let { defaultsObj.put("changes", toJSONArray(it)) }
            d?.selectedOptimizations?.let { defaultsObj.put("selectedOptimizations", toJSONArray(it)) }
            anonymousId?.let { defaultsObj.put("anonymousId", it) }
            if (defaultsObj.length() > 0) {
                obj.put("defaults", defaultsObj)
            }
        }

        return obj.toString()
    }

    private fun toJSONArray(list: List<Map<String, Any>>): org.json.JSONArray {
        val arr = org.json.JSONArray()
        for (item in list) {
            arr.put(JSONObject(item))
        }
        return arr
    }
}
